python/ray/data/_internal/block_batching.py

```diff
@@ -154,6 +158,16 @@ def batch_blocks(
         stats=stats,
     )
 
+    if collate_fn is not None:
+
+        def batch_fn_iter(
+            iterator: Iterator[DataBatch], collate_fn: Callable[[DataBatch], DataBatch]
+        ) -> Iterator[DataBatch]:
+            for batch in iterator:
+                yield collate_fn(batch)
+
+        batch_iter = batch_fn_iter(batch_iter, collate_fn)
```

Comment:
Nit: If going to use a dynamically defined function, you can just capture `collate_fn` in the closure.
```suggestion
        def batch_fn_iter(iterator: Iterator[DataBatch]) -> Iterator[DataBatch]:
            for batch in iterator:
                yield collate_fn(batch)

        batch_iter = batch_fn_iter(batch_iter)
```

Reply:
updated!